### PR TITLE
add enrollments list to targeting attributes for nimbus

### PR DESCRIPTION
**Pull request by jeddai**
https://mozilla-hub.atlassian.net/browse/EXP-3599

### Pull Request checklist ###

- **Breaking changes**:  This PR follows our [breaking change policy](https://github.com/mozilla/application-services/blob/main/docs/howtos/breaking-changes.md)
  - [ ] This PR follows the breaking change policy:
     - This PR has no breaking API changes, or
     - There are corresponding PRs for our consumer applications that resolve the breaking changes and have been approved
- [x] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes a changelog entry in [CHANGELOG.md](../CHANGELOG.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [x] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

[Branch builds](https://github.com/mozilla/application-services/blob/main/docs/howtos/branch-builds.md): add `[firefox-android: branch-name]` to the PR title.

### CODE REVIEW

File: components/nimbus/src/enrollment.rs [hunk 1]
```diff
@@ -28,7 +28,7 @@ cfg_if::cfg_if! {
 
 #[cfg_attr(not(feature = "stateful"), allow(unused))]
 const DEFAULT_GLOBAL_USER_PARTICIPATION: bool = true;
-pub(crate) const PREVIOUS_ENROLLMENTS_GC_TIME: Duration = Duration::from_secs(30 * 24 * 3600);
+pub(crate) const PREVIOUS_ENROLLMENTS_GC_TIME: Duration = Duration::from_secs(365 * 24 * 3600);
```
Comment (jeddai) [inline components/nimbus/src/enrollment.rs:31]: This change is covered in existing tests — the tests use the variable so they didn't need to be updated.

File: components/nimbus/src/tests/stateful/test_nimbus.rs [hunk 2]
```diff
@@ -971,3 +971,273 @@ fn test_fetch_enabled() -> Result<()> {
...
+    client.set_experiments_locally(experiment_json)?;
+    client.apply_pending_experiments()?;
+
+    let active_experiments = client.get_active_experiments()?;
+    assert_eq!(active_experiments.len(), 1);
+
+    let targeting_helper = client.create_targeting_helper(None)?;
+    assert!(!targeting_helper.eval_jexl("'test-1' in active_experiments".to_string())?);
+    assert!(targeting_helper.eval_jexl("'test-2' in active_experiments".to_string())?);
+    assert!(targeting_helper.eval_jexl("'test-1' in enrollments".to_string())?);
```
Comment (jaredlockhart) [inline components/nimbus/src/tests/stateful/test_nimbus.rs:1099]: So at this point we've unenrolled from `test-1` and enrolled in `test-2`, so shouldn't both `test-1` and `test-2` appear in `enrollments`?

Comment (jeddai) [reply components/nimbus/src/tests/stateful/test_nimbus.rs:1099]: Ah good call I'll add that there too

File: components/nimbus/src/tests/stateful/test_nimbus.rs [hunk 3]
```diff
@@ -971,3 +971,275 @@ fn test_fetch_enabled() -> Result<()> {
...
+            "targeting": "true",
+            "userFacingName": "test experiment",
+            "referenceBranch": "control",
+            "isEnrollmentPaused": false,
+            "proposedEnrollment": 7,
+            "userFacingDescription": "This is a test experiment for testing purposes.",
+            "id": "secure-copper",
+            "last_modified": 1_602_197_324_372i64,
+        }
+    ]}))?;
```
Comment (jhugman) [inline components/nimbus/src/tests/stateful/test_nimbus.rs:1041]: Can we put this in a helper function, and parameterize?

File: CHANGELOG.md [hunk 4]
```diff
@@ -6,6 +6,12 @@
 - Android: The JVM compatibility target is now version 17 ([#5651](https://github.com/mozilla/application-services/pull/5651))
   - _NOTE: This is technically a breaking change, but all existing downstream projects have already made the necessary changes._
 
+## Nimbus ⛅️🔬🔭
+
+### 🦊 What's Changed 🦊
+
+- Add `enrollments` value to `TargetingAttributes` (Nimbus only) — it is a set of strings containing all enrollments, past and present ([#5685](https://github.com/mozilla/application-services/pull/5685)).
```
Comment (jhugman) [inline CHANGELOG.md:13]: The `Nimbus only` is a little esoteric; perhaps you can refer to the stateful Nimbus SDK or the Nimbus SDK for mobile?

File: components/nimbus/src/nimbus_client.rs [hunk 5]
```diff
@@ -308,14 +308,20 @@ impl NimbusClient {
...
+        let mut is_enrolled_set = HashSet::<String>::new();
+        let mut all_enrolled_set = HashSet::<String>::new();
         for ee in prev_enrollments {
             if let EnrollmentStatus::Enrolled { .. } = ee.status {
-                set.insert(ee.slug.clone());
+                is_enrolled_set.insert(ee.slug.clone());
+                all_enrolled_set.insert(ee.slug.clone());
+            }
+            if let EnrollmentStatus::WasEnrolled { .. } = ee.status {
+                all_enrolled_set.insert(ee.slug.clone());
```
Comment (jhugman) [inline components/nimbus/src/nimbus_client.rs:319]: I'm surprised clippy didn't want you to use a match here. You probably should put an `else` in at least.
```suggestion
                is_enrolled_set.insert(ee.slug.clone());
                all_enrolled_set.insert(ee.slug.clone());
            } else if let EnrollmentStatus::WasEnrolled { .. } = ee.status {
                all_enrolled_set.insert(ee.slug.clone());
```

**Review by jhugman — APPROVED** [review]
The code is exactly how I saw this working.

A few comments, but good job.

File: components/nimbus/src/nimbus_client.rs (hunk 5 above)
Comment (jeddai) [reply components/nimbus/src/nimbus_client.rs:319]: I think I like how a match looks better, good call

**Pull request merged**